File: C/ex18/bst_solution.c
// Binary Search Tree - Insert and Search (Exercise 18) - Solution
// Build a binary search tree (BST) and perform search operations.

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct Node {
    int value;
    struct Node *left;
    struct Node *right;
} Node;

Node* create_node(int value) {
    Node *node = (Node *)malloc(sizeof(Node));
    if (node == NULL) {
        return NULL;
    }
    node->value = value;
    node->left = NULL;
    node->right = NULL;
    return node;
}

Node* insert(Node *root, int value) {
    if (root == NULL) {
        return create_node(value);
    }
    
    if (value < root->value) {
        root->left = insert(root->left, value);
    } else if (value > root->value) {
        root->right = insert(root->right, value);
    }
    
    return root;
}

int search(Node *root, int value) {
    if (root == NULL) {
        return 0;
    }
    
    if (value == root->value) {
        return 1;
    } else if (value < root->value) {
        return search(root->left, value);
    } else {
        return search(root->right, value);
    }
}

void delete_tree(Node *root) {
    if (root == NULL) {
        return;
    }
    
    delete_tree(root->left);
    delete_tree(root->right);
    free(root);
}

int main() {
    Node *root = NULL;
    char command[50];
    int value;
    
    while (1) {
        if (scanf("%s", command) != 1) {
            break;
        }
        
        if (strcmp(command, "exit") == 0) {
            break;
        } else if (strcmp(command, "insert") == 0) {
            if (scanf("%d", &value) == 1) {
                root = insert(root, value);
            }
        } else if (strcmp(command, "search") == 0) {
            if (scanf("%d", &value) == 1) {
                if (search(root, value)) {
                    printf("found\n");
                } else {
                    printf("not found\n");
                }
            }
        }
    }
    
    delete_tree(root);
    
    return 0;
}

File: C#/ex17/bst_solution.cs
// Binary Search Tree - Insert and Search (Exercise 17) - Solution
// Build a binary search tree (BST) and perform search operations.

using System;

class Node {
    public int Value { get; set; }
    public Node? Left { get; set; }
    public Node? Right { get; set; }
    
    public Node(int value) {
        Value = value;
        Left = null;
        Right = null;
    }
}

class BST {
    private Node? root;
    
    public BST() {
        root = null;
    }
    
    public void Insert(int value) {
        if (root == null) {
            root = new Node(value);
        } else {
            InsertRecursive(root, value);
        }
    }
    
    private void InsertRecursive(Node node, int value) {
        if (value < node.Value) {
            if (node.Left == null) {
                node.Left = new Node(value);
            } else {
                InsertRecursive(node.Left, value);
            }
        } else if (value > node.Value) {
            if (node.Right == null) {
                node.Right = new Node(value);
            } else {
                InsertRecursive(node.Right, value);
            }
        }
    }
    
    public bool Search(int value) {
        return SearchRecursive(root, value);
    }
    
    private bool SearchRecursive(Node? node, int value) {
        if (node == null) {
            return false;
        }
        
        if (value == node.Value) {
            return true;
        } else if (value < node.Value) {
            return SearchRecursive(node.Left, value);
        } else {
            return SearchRecursive(node.Right, value);
        }
    }
}

class Program {
    static void Main() {
        BST bst = new BST();
        
        while (true) {
            string? line = Console.ReadLine();
            if (line == null) break;
            
            var parts = line.Split(new[]{' ', '\t'}, StringSplitOptions.RemoveEmptyEntries);
            if (parts.Length == 0) continue;
            
            string command = parts[0];
            
            if (command == "exit") {
                break;
            } else if (command == "insert" && parts.Length >= 2) {
                if (int.TryParse(parts[1], out int value)) {
                    bst.Insert(value);
                }
            } else if (command == "search" && parts.Length >= 2) {
                if (int.TryParse(parts[1], out int value)) {
                    if (bst.Search(value)) {
                        Console.WriteLine("found");
                    } else {
                        Console.WriteLine("not found");
                    }
                }
            }
        }
    }
}
